File: learn/math/big-number/big_number.go


import (
	"strconv"
)

type BigInt struct {
	digits []int
	base   int
}

func NewBigInt() *BigInt {
	return &BigInt{
		digits: make([]int, 0),
		base:   1e9,
	}
}

func NewBigIntFromInt64(v int64) *BigInt {
	bi := &BigInt{base: 1e9}
	if v == 0 {
		bi.digits = append(bi.digits, 0)
	}
	for v > 0 {
		bi.digits = append(bi.digits, int(v%int64(bi.base)))
		v /= int64(bi.base)
	}
	return bi
}

func NewBigIntFromString(s string) *BigInt {
	bi := &BigInt{base: 1e9}
	if len(s) == 0 {
		bi.digits = append(bi.digits, 0)
		return bi
	}
	for len(s)%9 != 0 {
		s = "0" + s
	}
	for i := 0; i < len(s); i += 9 {
		v, _ := strconv.Atoi(s[i : i+9])
		// prepend to digits
		bi.digits = append([]int{v}, bi.digits...)
	}
	bi.removeLeadingZeros()
	return bi
}

func (b *BigInt) removeLeadingZeros() {
	for len(b.digits) > 1 && b.digits[len(b.digits)-1] == 0 {
		b.digits = b.digits[:len(b.digits)-1]
	}
}

func (b *BigInt) Add(other *BigInt) *BigInt {
	res := NewBigInt()
	carry := 0
	maxLen := len(b.digits)
	if len(other.digits) > maxLen {
		maxLen = len(other.digits)
	}

	for i := 0; i < maxLen || carry > 0; i++ {
		if i < len(b.digits) {
			carry += b.digits[i]
		}
		if i < len(other.digits) {
			carry += other.digits[i]
		}
		res.digits = append(res.digits, carry%b.base)
		carry /= b.base
	}
	res.removeLeadingZeros()
	return res
}

func (b *BigInt) LessThan(other *BigInt) bool {
	if len(b.digits) != len(other.digits) {
		return len(b.digits) < len(other.digits)
	}
	for i := len(b.digits) - 1; i >= 0; i-- {
		if b.digits[i] != other.digits[i] {
			return b.digits[i] < other.digits[i]
		}
	}
	return false
}

func (b *BigInt) Equals(other *BigInt) bool {
	return !b.LessThan(other) && !other.LessThan(b)
}

File: learn/math/big-number/big_number.rs
use std::cmp::Ordering;

#[derive(Clone, Eq, PartialEq)]
pub struct BigInt {
    digits: Vec<i32>,
    base: i32,
}

impl BigInt {
    pub fn new() -> Self {
        BigInt {
            digits: Vec::new(),
            base: 1_000_000_000,
        }
    }

    pub fn from_i64(mut v: i64) -> Self {
        let mut bi = BigInt::new();
        if v == 0 {
            bi.digits.push(0);
        }
        let base_i64 = bi.base as i64;
        while v > 0 {
            bi.digits.push((v % base_i64) as i32);
            v /= base_i64;
        }
        bi
    }

    pub fn from_string(mut s: String) -> Self {
        let mut bi = BigInt::new();
        if s.is_empty() {
            bi.digits.push(0);
            return bi;
        }
        while s.len() % 9 != 0 {
            s.insert(0, '0');
        }
        for i in (0..s.len()).step_by(9) {
            let v = s[i..i + 9].parse::<i32>().unwrap_or(0);
            bi.digits.insert(0, v);
        }
        bi.remove_leading_zeros();
        bi
    }

    fn remove_leading_zeros(&mut self) {
        while self.digits.len() > 1 && *self.digits.last().unwrap() == 0 {
            self.digits.pop();
        }
    }

    pub fn add(&self, other: &Self) -> Self {
        let mut res = BigInt::new();
        let mut carry = 0;
        let mut i = 0;

        while i < self.digits.len() || i < other.digits.len() || carry > 0 {
            if i < self.digits.len() {
                carry += self.digits[i];
            }
            if i < other.digits.len() {
                carry += other.digits[i];
            }
            res.digits.push(carry % self.base);
            carry /= self.base;
            i += 1;
        }
        res.remove_leading_zeros();
        res
    }
}

impl PartialOrd for BigInt {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for BigInt {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.digits.len() != other.digits.len() {
            return self.digits.len().cmp(&other.digits.len());
        }
        for i in (0..self.digits.len()).rev() {
            if self.digits[i] != other.digits[i] {
                return self.digits[i].cmp(&other.digits[i]);
            }
        }
        Ordering::Equal
    }
}
